processor/samplingprocessor/tailsamplingprocessor/processor.go

```diff
@@ -244,37 +241,49 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pdata.TraceID, trace *samp
 }
 
 // ConsumeTraceData is required by the SpanProcessor interface.
-func (tsp *tailSamplingSpanProcessor) ConsumeTraces(_ context.Context, td pdata.Traces) error {
+func (tsp *tailSamplingSpanProcessor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
 	tsp.start.Do(func() {
 		tsp.logger.Info("First trace data arrived, starting tail_sampling timers")
 		tsp.policyTicker.Start(1 * time.Second)
 	})
 
-	octds := internaldata.TraceDataToOC(td)
-	var errs []error
-	for _, octd := range octds {
-		if err := tsp.processTraces(octd); err != nil {
-			errs = append(errs, err)
+	resourceSpans := td.ResourceSpans()
+	sampledTraceData := pdata.NewTraces()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		resourceSpan := resourceSpans.At(i)
+		if resourceSpan.IsNil() {
+			continue
 		}
+		tsp.processTraces(ctx, resourceSpan, sampledTraceData)
 	}
-	return componenterror.CombineErrors(errs)
+	return nil
 }
 
-func (tsp *tailSamplingSpanProcessor) processTraces(td consumerdata.TraceData) error {
-
-	// Groupd spans per their traceId to minimize contention on idToTrace
-	idToSpans := make(map[traceKey][]*tracepb.Span)
-	for _, span := range td.Spans {
-		if len(span.TraceId) != 16 {
-			tsp.logger.Warn("Span without valid TraceId", zap.String("SourceFormat", td.SourceFormat))
+func (tsp *tailSamplingSpanProcessor) groupSpansByTraceKey(resourceSpans pdata.ResourceSpans, td pdata.Traces) map[traceKey][]*pdata.Span {
+	idToSpans := make(map[traceKey][]*pdata.Span)
+	ilss := resourceSpans.InstrumentationLibrarySpans()
+	for j := 0; j < ilss.Len(); j++ {
+		ils := ilss.At(j)
+		if ils.IsNil() {
 			continue
 		}
-		tk := traceKey(span.TraceId)
```

Comment:
Is this type (`traceKey`) still needed? Can't we have the `pdata.TraceID` as the key? From what I understand, this is converting the `[]byte` representation of the `TraceID` to a string, only to have it as key to the map.

Reply:
correct. let me give it a try...

Reply:
runtime error: hash of unhashable type pdata.TraceID : Cant use TraceID as the key.

Reply:
Was worth the attempt :-)

Reply:
yup